mne_sandbox/preprocessing/_sns.py

```diff
@@ -0,0 +1,168 @@
+            raise ValueError('n_neighbors must be at most len(good_picks) '
+                             '- 1 (%s)' % (len(good_picks) - 1,))
+        logger.info('    Loading data')
+        raw = raw.copy()
+        picks = _pick_data_channels(raw.info, exclude=())
+        # The following lines are equivalent to this, but require less mem use:
+        # data_cov = np.cov(orig_data)
+        # data_corrs = np.corrcoef(orig_data) ** 2
+        logger.info('    Computing covariance for %s good channels'
+                    % len(good_picks))
+        data_cov = np.eye(len(picks))
+        good_cov = compute_raw_covariance(
+            raw, picks=good_picks, reject=reject, flat=flat,
+            verbose=False)['data']
+        # re-index this
+        good_picks = np.searchsorted(picks, good_picks)
+        bad_picks = np.setdiff1d(np.arange(len(picks)), good_picks)
+        data_cov[np.ix_(good_picks, good_picks)] = good_cov
+        del good_picks
+        data_norm = np.diag(data_cov)
+        data_corrs = data_cov * data_cov
+        data_corrs /= data_norm
+        data_corrs /= data_norm[:, np.newaxis]
+        del data_norm
+        data_cov *= len(raw.times)
+        operator = np.zeros((len(picks), len(picks)))
+        logger.info('    Assembling spatial operator')
+        for ii in range(len(picks)):
+            # For each channel, the set of other signals is orthogonalized by
+            # applying PCA to obtain an orthogonal basis of the subspace
+            # spanned by the other channels.
+            if ii in bad_picks:
+                operator[ii, ii] = 1.
+                continue
+            idx = np.argsort(data_corrs[ii])[::-1][:n_neighbors + 1].tolist()
+            idx.pop(idx.index(ii))  # should be in there
+            # XXX Eventually we might want to actually threshold here (with
+            # rank-deficient data it could matter)
+            eigval, eigvec = _pca(data_cov[np.ix_(idx, idx)], thresh=None)
+            eigvec *= 1. / np.sqrt(eigval)
+            del eigval
+            # augment with given channel
+            eigvec = np.vstack(([[1] + [0] * n_neighbors],
+                               np.hstack((np.zeros((n_neighbors, 1)),
+                                          eigvec))))
+            idx = np.concatenate(([ii], idx))
+            corr = np.dot(np.dot(eigvec.T, data_cov[np.ix_(idx, idx)]), eigvec)
+            # The channel is projected on this basis and replaced by its
+            # projection
+            operator[ii, idx[1:]] = np.dot(corr[0, 1:], eigvec[1:, 1:].T)
+        logger.info('Done')
+        self._operator = operator
+        self._used_chs = [raw.ch_names[pick] for pick in picks]
+
+    @property
+    def operator(self):
+        """The operator matrix"""
+        return self._operator.copy()
+
+    def apply(self, inst):
+        """Apply the operator
+
+        Parameters
+        ----------
+        inst : instance of Raw
```

Comment:
see code below with XXX comment, eventually it could be other instances (and probably would need to be that way to go into `mne` proper)